Optimizer: Re-use CFG from type inference

topolarity opened the pull request:
This change will allow us to re-use Inference-collected information at the basic block level, such as the `bb_vartables`.

There were a couple of spots where the `unreachable` insertion pass at IRCode conversion (i.e. optimizer entry) was ignoring statically divergent code that inference had discovered:
  - `%x = SlotNumber(3)` can throw and cause the following statements to be statically unreachable
  - `goto #b if not %cond` can be statically throwing if %cond is known to never be Bool (or to always throw during its own evaluation)

CFG re-computation was hiding these bugs by flowing through the "Core.Const(...)"-wrapped statements that would follow, inserting unnecessary but harmless extra branches in the CFG.

topolarity committed:
Don't re-compute CFG on Optimizer entry

This change will allow us to re-use Inference-collected information at
the basic block level, such as the `bb_vartables`.

There were a couple of spots where the `unreachable` insertion pass at
IRCode conversion (i.e. optimizer entry) was ignoring statically
divergent code that inference had discovered:
  - `%x = SlotNumber(3)` can throw and cause the following statements to
    be statically unreachable
  - `goto #b if not %cond` can be statically throwing if %cond is known
    to never be Bool (or to always throw during its own evaluation)

CFG re-computation was hiding these bugs by flowing through the
"Core.Const(...)"-wrapped statements that would follow, inserting
unnecessary but harmless extra branches in the CFG.

topolarity committed:
IRVerify: Check IR stmt + metadata vector ranges

These checks are important to make sure that `verify_ir` does not end up
triggering out-of-bounds errors as it checks IR.

It also fills out some of the verification:
  - Confirms that cfg.index is correct
  - Confirms that cfg.blocks[i].stmts cover the complete IR

gbaraldi commented:
Potentially we could call adopt_thread on these threads but I would hold off on that until we figure some of the weird behaviour around them.

vchuravy commented on src/timing.c:
```diff
@@ -342,6 +342,7 @@ JL_DLLEXPORT void _jl_timing_block_init(char *buf, size_t size, jl_timing_event_
 JL_DLLEXPORT void _jl_timing_block_start(jl_timing_block_t *block) {
     assert(!block->is_running);
     if (!_jl_timing_enabled(block->event->subsystem)) return;
+    if (jl_get_pgcstack() == NULL) return; // not setup on this thread
```
Separate this change out?

Keno commented on base/compiler/optimize.jl:
```diff
@@ -555,54 +555,93 @@ function convert_to_ircode(ci::CodeInfo, sv::OptimizationState)
+                    if blockchangemap === nothing
+                        blockchangemap = fill(0, length(sv.cfg.blocks))
+                    end
+                    blockchangemap[block] += 1
```
Could you elaborate why it is required to create new blocks here?

Keno replied:
Oh wait, I misunderstood what this does. It's just for truncating the block. That's fine.

topolarity committed:
test: allow `nothing; return` in `IRUtils.fully_eliminated`

For some reason the recent CFG change caused us to generate:
```
1-element Vector{Any}:
 CodeInfo(
1 ─     nothing::Nothing
└──     return true
) => Bool
```

instead of

```
1-element Vector{Any}:
 CodeInfo(
1 ─     return true
) => Bool
```

but I believe this difference is harmless.

topolarity commented on base/compiler/optimize.jl:
```diff
@@ -847,3 +886,17 @@ function renumber_ir_elements!(body::Vector{Any}, ssachangemap::Vector{Int}, lab
     end
 end
+
+function renumber_cfg_blocks!(cfg::CFG, blockchangemap::Vector{Int})
```
Yeah good point, I agree

aviatesk commented on base/compiler/typeinfer.jl:
```diff
@@ -551,9 +551,9 @@ function finish(me::InferenceState, interp::AbstractInterpreter)
         # either because we are the outermost code, or we might use this later
         doopt = (me.cached || me.parent !== nothing)
-        recompute_cfg = type_annotate!(interp, me, doopt)
+        type_annotate!(interp, me, doopt)
```
We can eliminate anything related to `any_unreachable` within `type_annotate!` now.

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -555,54 +555,93 @@ function convert_to_ircode(ci::CodeInfo, sv::OptimizationState)
             prevloc = codeloc
         end
-        if code[idx] isa Expr && ssavaluetypes[idx] === Union{}
+        if ssavaluetypes[idx] === Union{} && !(code[idx] isa Core.Const)
```
Maybe we want to set up a special and explicit marker type instead of `Const` to indicate a statement is known to be unreachable, e.g.
```julia
struct Unreachable
    stmt
    Unreachable(@nospecialize stmt) = new(stmt)
end

aviatesk reviewed (commented):
Looks very nice. Sorry for comments after merge, but we can address them later anyway.
